### ISSUE TITLE
Update záznamu ve zdrojovém katalogu - odkazy do Krameria (primárně Aleph)

### ISSUE THREAD

**Issue by ZdenkaSera** [proarc/proarc#1706]
Možnost zapsat odkaz do zdrojového katalogu odesláním z ProArcu
- domluvit základ společný pro všechny (tj. podpole 856)
- vyjasnit pravidla, z jaké úrovně objektu odkaz tvořit/odesílat
- konfigurovatelný prefix odkazu do Krameria
- konfigurovatelný text (na instanci ProArcu)  pro podpole 856_y
- konfigurovatelná další případná pole k zapsání do katalogu (např. 911 pro odeslání do SKC; KNAV využívá pole 599 k označení záznamů, ze kterých se má odkaz poslat do registru digitalizace ...)

V KNAV probíhá analýza možností API, které poskytuje Aleph.

**Comment by SykoraLukas** [comment]
Požadavek MZK je, aby se exportoval při každém exportu csv soubor do specialni cesty, odkud probiha sklizeni Alephem. Požadavky KNAV budou řešeny následně i s jejich oddělením, které se stará o katalogy.

Export je konfigurovatelný, defaultně vypnutý.

**Comment by SykoraLukas** [comment]
[Nutná konfigurace ](https://github.com/proarc/proarc/wiki/Konfigurace#aleph-xml-katalog)

**Comment by ZdenkaSera** [comment]
Schůzka 3.9.2024:
-  Pro Aleph bude zapotřebí plochý soubor (csv, txt), který bude obsahovat buď a) jen údaje specifické pro konkrétní objekt (zřejmě sysno + odkaz) a ostatní (statické) údaje vygeneruje administrátor alephu skriptem nebo b) všechny informace k zápisu do zýáznamu ve formátu již připraveném pro zápis
- V diskutovaném návrhu chybí řešení pro zápis údajů o digitalizovaných částech periodik k následnému odeslání do SKC - @kerschfilip doplní

**Comment by SykoraLukas** [comment]
Doplněno do API i klienta.

![Image](https://github.com/user-attachments/assets/856eb53c-30ab-443c-b90c-50dd263ec919)

![Image](https://github.com/user-attachments/assets/cade1af2-81c1-4853-bea5-3c2a18b1a708)

[40482007-aa54-4f97-9781-ae8f381d4fb0.csv](https://github.com/user-attachments/files/17342486/40482007-aa54-4f97-9781-ae8f381d4fb0.csv)


Do konfigurace API se musi doplnit/upravit [katalog](https://github.com/proarc/proarc/wiki/Konfigurace#aleph-xml-katalog), proti kteremu se validuji data

```properties
# Aleph Knihovna AV ČR
catalog.knaValidation.url=****
catalog.knaValidation.base=***
catalog.knaValidation.name=Knihovna AV Validace
catalog.knaValidation.type=Z3950Catalog
catalog.knaValidation.recordCharset=UTF-8
catalog.knaValidation.fields=sys
catalog.knaValidation.field.sys.query=@attrset bib-1 @attr 1=12 @attr 4=1 "%s"
catalog.knaValidation.field.sys.title=SYS
catalog.knaValidation.updateType=ALEPH_XML
catalog.knaValidation.catalogDirectory=/data/aleph/
catalog.knaValidation.catalogUrlLink=https://kramerius.lib.cas.cz/view/
catalog.knaValidation.sysnoLenght=-1
catalog.knaValidation.checkSysnoBeforeUpdate=true
catalog.knaValidation.defaultSearchField=sys
```

A dopolnit nova property [validationCatalogs](https://github.com/proarc/proarc/wiki/Konfigurace#katalogy-s-metadaty), ktera poskytuje katalogy, proto kterym uzivatele smi validovat data. Katalogy totiž mohou byt odlisne oproti těm, ze kterych se data stahují.

``` properties
## Bibliographic metadata catalogs, that allow update record
validationCatalogs=knaValidation
```

Na straně klieta je potřeba přidat propertu, která zaktivní funkci updatu zaznamu ve zdrojovem katalogu a definovat seznam modelů, nad kterými toto lze spustit.
``` properties
"updateInSource": true,
"updateInSourceModels": ["model:ndkperiodical", "model:ndkmonographvolume", "model:ndksheetmusic"]
```

**Comment by SykoraLukas** [comment]
K vyzkoušení na Inovatice.

**Comment by ZdenkaSera** [comment]
Test na masteru, úplný text až lokálně.
- [ ] Fce je na masteru k dispozici a lze ji spustit - pouze pro superadmina
- [ ] Objeví se ve správě procesů mezi interními procesy
- [ ] Chybí label k profilu - formField.profiles.internalProfile.updateCatalogRecords
- [ ] Chybí v seznamu profilů k filtraci
- [ ] Lze ji spustit i nad objektem, který nebyl vůbec exportován - zvážit kontrolu (exp. K nebo N)?
- [ ] Zvážit oprávnění ?

**Comment by ZdenkaSera** [comment]
Komentář výše upraven po schůzce 18.11.2024.

**Comment by SykoraLukas** [comment]
> * [ ]  Lze ji spustit i nad objektem, který nebyl vůbec exportován případně nemá strany - zvážit kontrolu (proběhl exp. K nebo N)?

Toto si musí říct uživatelé, jak mají nastavené procesy. Podle mě není nutnost kontrolovat export K nebo N.

**Comment by SykoraLukas** [comment]
Kontrola na oprávnění doplněna viz https://github.com/proarc/proarc-client/issues/460#issuecomment-2505518277 spolu s ostatními body, které se týkají klienta.

**Comment by SykoraLukas** [comment]
Po úspěšném automatickém exportu do Krameria se může automaticky zapsat i do katalogu. Pro to je potřeba u instance Krameria nastavit katalog, ktery se bude kontrolovat a editovat.

```
# (4.2.1) id katalogu, kam se ma provest upload informaci o zdigitalizovani. Pokud neni vyplněné, nikam se informace neposílají
# Pokud se po úspěšném automatickém exportu do Krameria má zapsat změna do zdrojového katalogu, je nutné nastavit, jaký do jakého katalogu se změna propisuje. Do property je nutné zapsat id katalogu. Zároveň je v konfiguraci katalogu nutné definovat, jak má probíhat upload dat do katalogu. V současné době verze 4.2.1 je podporována změna do katalogu Verbis (pomocí volání API Verbisu) a do katalogu Aleph (pomocí vytvoření XML souboru).
krameriusInstance.kramerius5.uploadToCatalog=oai
krameriusInstance.kramerius7.uploadToCatalog=aleph_nkp
```

**Comment by SykoraLukas** [comment]
Zatim odebírám label chyba, je pripraveno k testovani.

**Comment by ZdenkaSera** [comment]
@SykoraLukas Otestovali jsme ruční odesílání csv i jejich tvorbu při automatickém importu do Krameria a zatím to vypadá moc pěkně. Pokud bychom na něco nečekaného narazili v ostrém provozu, založíme issue.

Tady bych chtěla otevřít jedinou věc: Využíváme v nastavení kontrolu délky sysna - pokud je delší než zadaný parametr, odkaz se při ručním odeslání nevygeneruje - to je správně. 

1) Kosmetika - nešlo by vylepšit s tím spojenou hlášku, aby uživatel pochopil důvod?

![Image](https://github.com/user-attachments/assets/7ff8cc69-cac9-4be5-bf15-5b98cd7b02d0)

2) Chyba - při automatickém importu do Krameria tato situace nedovolí vůbec uskutečnit export. Z našeho pohledu je to ale situace analogická té, kdy např. exportujeme objekt, jehož sysno sice má správnou délku, ale nenajde se v našem katalogu (typicky: externí zakázky spojené se zveřejněním v našem Krameriovi). Prosím, podíváš se na to? Děkuji.

**Comment by SykoraLukas** [comment]
> 2\. Chyba - při automatickém importu do Krameria tato situace nedovolí vůbec uskutečnit export. Z našeho pohledu je to ale situace analogická té, kdy např. exportujeme objekt, jehož sysno sice má správnou délku, ale nenajde se v našem katalogu (typicky: externí zakázky spojené se zveřejněním v našem Krameriovi). Prosím, podíváš se na to? Děkuji.

Jaká jsi zjitila, ze to nedovolí import do Krameria?

**Comment by ZdenkaSera** [comment]
Ještě doplňuji, že autoimporty s iinými "chybami sysna" fungují bez problémů, ve správě o importu pak jsou srozumitelné hlášky:

Import do Krameria proběhl, ale nepodařilo se upravit záznam v katalogu.Záznam v katalogu (kna, sysno: 000960979) již obsahuje info o digitalizaci.

Import do Krameria proběhl, ale nepodařilo se upravit záznam v katalogu.msg: Nenalezena žádná data v katalogu kna pro identifikator 00231016X

**Comment by SykoraLukas** [comment]
Můžu poprosit o zaslání konfigurace proarcu? Stačí kompletní sekce katalogu a nakonfigurovaných instatncí Krameria (bez chráněných údajů).

**Comment by SykoraLukas** [comment]
Opraveno, bude v dalsi verzi.

**Comment by ZdenkaSera** [comment]
@SykoraLukas Prosím, byl by možný rychlý fix vydané verze? Opravu bych otestovala na masteru. Předem díky.

Ráno jsem lokálně otestovala jen sysno delší než 9, to bylo ok, ale stejnou chybu PA vrací i tam, kde by měl vytvořit csv.. 
Test na masteru na objektu uuid:5e1f9584-699e-41d3-b73f-065a34c382a5 - tady by se odkaz měl vygenerovat.

Updatovano 0 z 1
uuid:5e1f9584-699e-41d3-b73f-065a34c382a5 Catalog with id knaValidation does not support Record modification

**Comment by SykoraLukas** [comment]
Opraveno.

**Comment by ZdenkaSera** [comment]
Díky, ale ještě to není ono. Prosím, podíváš se ještě? Děkuji.
- uuid:39c7d6b0-2095-4211-9fe2-56994fa54b08 - sysno s délkou 10 (započítávají se do ní vedoucí 0) a v KNAV neexistuje - podle správy procesů se odkaz vytvoří, měla být chyba

Ostatní varianty byly ok:
- uuid:5e1f9584-699e-41d3-b73f-065a34c382a5  - odkaz se má zapsat - to je ok
- uuid:adcc18e7-28ab-41c8-a277-a7ba681d9c22 - neobsahuje strany - hlásí chybu - je ok
- uuid:24bba0e3-e886-4286-a9f1-3c003837b6b9 - už má odkaz, ale s jiným uuid - je ok
- uuid:86562d16-27cf-49c8-b133-56826894404f - má odkaz v katalogu se stejným uuid - hlásí chybu - je ok
- uuid:db7f7738-bcb3-4862-a42e-e082d4c6c353 - sysno s délkou 9, není v KNAV - hlásí chybu - je ok

**Comment by SykoraLukas** [comment]
Zminovany zaznam ma sysno ve vasem katalogu, proto se zaznam pro katalog zapise.

Obsah vytvořeného souboru:
`0018366761 @ https://kramerius.lib.cas.cz/view/uuid:39c7d6b0-2095-4211-9fe2-56994fa54b08`


Pokus o vyhledání sysna:

![Image](https://github.com/user-attachments/assets/6568575d-00cf-4373-ab69-4205c70cf5b9)

**Comment by SykoraLukas** [comment]
Ono to bude mozna souviset nejak s nastavenim zpracovani dotazu u vaseho katalogu. Hledam size 10 znakove sysno, ale v odpovedi, co dostavam zpet ma sysno jen 9 znaku (chybi posledni znak).

**Comment by ZdenkaSera** [comment]
Koukám na to - když to nechám vyhledat v prohlížeči (vufindu), nenajde to nic. Když jsem se přihlásila přímo do alephu, našlo to záznam bez koncové jedničky (jako na tvém obrázku). To je ale malér, nenapadá mě. co s tím. Prosím tě, nešlo by vyjít z nastavení délky sysna v konfigu katalogu a pokud bude delší, rovnou vydat chybu a proti katalogu to neověřovat?

**Comment by SykoraLukas** [comment]
Ano, tuto kontrolu tam uz mame. Jen se to musi nastavit v konfigu daneho katalogu. Na jakou delku to mam nastavit? Pokud tam je hodnota -1 (kontrola na delku se neprovádí).  Což je náš případ. Mám tedy nastavit kontrolu na délku 9?

**Comment by SykoraLukas** [comment]
Kontrola nastavena na délku 9 znaků.  Odpověď ProArcu poté je:

![Image](https://github.com/user-attachments/assets/5b4b8680-1694-49c0-acc2-4d11707f2823)

**Comment by SykoraLukas** [comment]
Nastaven autoimport.

**Comment by ZdenkaSera** [comment]
Moc děkuju, ruční odesílání csv je ok, nenašla jsem protipříklad 😊

**Comment by ZdenkaSera** [comment]
Lukáši, ten automimport mi nabízí je lokál - můžu ještě prosit ...?

![Image](https://github.com/user-attachments/assets/8775641b-4e3c-4f28-a6e0-fb929b3fe9d9)

**Comment by SykoraLukas** [comment]
Pod jakym uzivatelem? Muzes si nastavit opravneni na "Import do produkcniho Krameria"?

**Comment by luckajirku** [comment]
a není to zas tenhle problém https://github.com/proarc/proarc/issues/1034 ?

**Comment by ZdenkaSera** [comment]
Lukáši, import selhal - nebudu tě týrat, otestuju ty importy až lokálně. Moc díky!

**Comment by SykoraLukas** [comment]
Proces pada na Krameriovi, kluci tam meni to akubra jadro a asi to jeste neni zcela odladene. 

Chyba z Krameria:

![Image](https://github.com/user-attachments/assets/bade7a66-3951-4681-99c7-05916ae24742)

**Comment by ZdenkaSera** [comment]
Ano, děkuju - tuším, ve správě to říkalo, že selhal import, takže na odkazy ani nedošlo :-)

**Comment by ZdenkaSera** [comment]
@luckajirku Díky za připomenutí, myslím, že to není úplně totéž a že to nemusíme řešit (tedy alespoň ne v rámci KNAV).  Předávám to ale @tumovam

**Comment by ZdenkaSera** [comment]
Tvorba odkazů (ruční i po autoimportu do K7) ve verzi 4.2.9/2.3.5 vypadá dobře. V této verzi testováno jen pro NDK monografie, pro jiné objekty zatím generování odkazů nevyužíváme.t
